File: Data Structures & Algorithms/course-schedule/submission-0.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:

        ad_list = defaultdict(list)
        
        for course, pre in prerequisites:
            ad_list[course].append(pre)
        
        def dfs(ad_list, visited, i):
            if len(ad_list[i]) == 0:
                return True

            if i in visited:
                return False
            
            visited.add(i)

            for pre in ad_list[i]:
                if not dfs(ad_list, visited, pre):
                    return False

            visited.remove(i)
            return True


        for i in range(numCourses):
            visited = set()
            if not dfs(ad_list, visited, i):
                return False

        return True

        

File: Data Structures & Algorithms/course-schedule-ii/submission-0.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> adList = new HashMap<>();

        for(int i = 0; i < numCourses; i++) {
            adList.put(i, new ArrayList<>());
        }

        for(int[] prereq : prerequisites) {
            List<Integer> coursePres = adList.get(prereq[0]);
            coursePres.add(prereq[1]);
        }

        List<Integer> outputList = new ArrayList<>();
        Set<Integer> visited = new HashSet<>();
        Set<Integer> cycle = new HashSet<>();
        for(int i = 0; i < numCourses; i++) {
            if(!dfs(adList, visited, cycle, i, outputList)) {
                return new int[0];
            }
        }

        int[] output = new int[outputList.size()];
        for(int i = 0; i < outputList.size(); i++) {
            output[i] = outputList.get(i);
        }
        return output;
    }

    public boolean dfs(Map<Integer, List<Integer>> adList, Set<Integer> visited, Set<Integer> cycle, int node, List<Integer> outputList) {
        if(cycle.contains(node)) {
            return false;
        }
        if(visited.contains(node)) {
            return true;
        }

        cycle.add(node);
        List<Integer> pres = adList.get(node);
        for(int pre : pres) {
            if(!dfs(adList, visited, cycle, pre, outputList)) {
                return false;
            }
        }
        cycle.remove(node);
        visited.add(node);
        outputList.add(node);

        return true;
    }
}
